src/write.jl

```diff
@@ -206,73 +200,59 @@ function xlsxescape(str::AbstractString)
     return replace(str, "'", "&apos;")
 end
 
-function setdata!(ws::Worksheet, ref::CellRef, str::AbstractString)
-    if isempty(str)
-        # no-op
-        return
+"""
+Returns the datatype and value for `val` to be inserted into `ws`.
+"""
+function xlsx_encode(ws::Worksheet, val::AbstractString)
+    if isempty(val)
+        return "", ""
     end
-
-    sst_ind = add_shared_string!(get_workbook(ws), xlsxescape(str))
-    cell = Cell(ref, "s", "", string(sst_ind), "")
-    setdata!(ws, cell)
+    sst_ind = add_shared_string!(get_workbook(ws), xlsxescape(val))
+    return "s", string(sst_ind)
 end
 
-function setdata!(ws::Worksheet, ref::CellRef, m::Missings.Missing)
-    # no-op
-    nothing
-end
+xlsx_encode(::Worksheet, val::Missings.Missing) = ("", "")
+xlsx_encode(::Worksheet, val::Bool) = ("b", val ? "1" : "0")
+xlsx_encode(::Worksheet, val::Union{Int, Float64}) = ("", string(val))
+xlsx_encode(ws::Worksheet, val::Date) = ("", string(date_to_excel_value(val, isdate1904(get_xlsxfile(ws)))))
+xlsx_encode(ws::Worksheet, val::Dates.DateTime) = ("", string(datetime_to_excel_value(val, isdate1904(get_xlsxfile(ws)))))
+xlsx_encode(::Worksheet, val::Dates.Time) = ("", string(time_to_excel_value(val)))
 
-function setdata!(ws::Worksheet, ref::CellRef, b::Bool)
-    value_str = b ? "1" : "0"
-    cell = Cell(ref, "b", "", value_str, "")
-    setdata!(ws, cell)
-end
-
-const DEFAULT_DATE_numFmtId = 14
-const DEFAULT_TIME_numFmtId = 20
-const DEFAULT_DATETIME_numFmtId = 22
+function get_style_index(ws::Worksheet, formatid::Integer)
+    @assert formatid >= 0
```

Comment:
Should you just make a `CellDataFormat` here?

Reply:
No, its a numFmtId. I'll probably replace it with a type but I want to leave that for when I add a bunch of other fomatting types and do a single wave of deprecations.